src/pkcs11/slot.c

```diff
@@ -427,6 +428,17 @@ card_detect_all(void)
 				}
 			}
 			card_detect(reader);
+
+			/* If reader was removed, card_removed() was already called in card_detect(),
+			 * remove only relation between reader and slot. */
+			if (reader->flags & SC_READER_REMOVED) {
```

Comment:
I assume this special case would happen if the reader was removed while being in this very `else` branch. However, is this special treatment really needed? I feel this is not a relevant issue, because it is very unlikely to happen at all *and* the bad consequences are unproblematic (i.e. some slot that was already removed may be visible to the application until card_detect_all is called the next time).

Reply:
It doesn't only cover the case, when reader is removed during the `else` branch,. The `card_detect_all()` does not update the the state of readers when looping over them:

https://github.com/OpenSC/OpenSC/blob/f7070f4c9f9c6e1cb8d2efe838f73d54c78553da/src/pkcs11/slot.c#L393-L398

So it catches the case, when reader is removed _after_ calling `refresh_attributes()`, so `reader->flags` are not yet updated.

Reply:
If that is your intention, then I think this code should not be added at all, because modifying the virtual slots would result in a different slotlist that are communicated to the application. that list, however needs to be static until the application calls C_GetSlotList, see https://github.com/OpenSC/OpenSC/issues/3241#issuecomment-2418834920

Reply:
I actually didn't realize this won't align with the C_GetSlotList behavior, thanks for pointing out. In that case, this PR will just add the correct check for `SC_READER_CARD_CHANGED` and will not touch the connection between slot and reader.